### DISCUSSION TITLE
Is there a reason why ArraySegment has to be one dimensional

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#91372] [General]
according to https://learn.microsoft.com/en-us/dotnet/api/system.arraysegment-1?view=net-7.0#remarks it is
>The original array must be one-dimensional and must have zero-based indexing

but why

**Comment by huoyaoyuan** [comment]
Because it uses 1d scalars to store index and length. For MD arrays it's not practical:

- Segmenting on total length is not logically meaningful. Dimensions can be teared.
- Segmenting for every dimension is not performant. The storage size has to be dynamic, and the indices has to be recalculated.

Indeed, SZ array and MD array are different types. There's no way to declare a variable work for them both.

**Reply by Xyncgas** [reply]
>Segmenting on total length is not logically meaningful. Dimensions can be teared.

why not, just segment on total length, get the nth element if that's what the user want and just count to n by enumerating through all elements in all dimension, why would this be useful you might ask, well if you are on 32 bit system but you can get more than 4GB of memory somehow, you now can store all these memory in array of array

**Reply by huoyaoyuan** [reply]
> well if you are on 32 bit system but you can get more than 4GB of memory somehow

I'm not aware of any way to achieve this in managed world. The *somehow* is totally fictional.

> get the nth element if that's what the user want and just count to n by enumerating through all elements in all dimension

This is https://github.com/dotnet/runtime/issues/49269#issuecomment-1112244449

> you now can store all these memory in array of array

The blocker is that all the linear memory abstractions are already using 32 bit length (). The total length of multi-dimensional array is the only exception (https://github.com/dotnet/runtime/issues/66775#issuecomment-1071427203).

`ArraySegment` is functionally replaced by `Span` or `Memory`, while all of them are not compatible for 64 bit storage at all.

**Reply by Xyncgas** [reply]
Thanks for the response, it was helpful

>I'm not aware of any way to achieve this in managed world. The somehow is totally fictional.

For example, you can get more than 4GB of memory in those system where a process is limited to 4GB memory by spawning multiple processes and use inter-process communication to share memory. It's complicated but definitely you are only limited by how much physical ram you got

**Reply by huoyaoyuan** [reply]
> use inter-process communication to share memory

Whatever accessing method you are using, you will not be able to access more than 4GB memory directly at a time. It's just limited by your address space. Only stream-like indirect access can satisfy that.
